src/png/scan_lines.rs

```diff
@@ -24,20 +24,24 @@ impl<'a> Iterator for ScanLines<'a> {
     type Item = ScanLine<'a>;
     #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        self.iter.next().map(|(len, pass, num_pixels)| {
-            let (data, rest) = self.raw_data.split_at(len);
-            self.raw_data = rest;
-            let (&filter, data) = if self.has_filter {
-                data.split_first().unwrap()
-            } else {
-                (&0, data)
-            };
-            ScanLine {
-                filter,
-                data,
-                pass,
-                num_pixels,
-            }
+        let (len, pass, num_pixels) = self.iter.next()?;
+        debug_assert!(self.raw_data.len() >= len);
+        debug_assert!(!self.has_filter || len > 1);
+        if self.raw_data.len() < len {
+            return None;
+        }
```

Comment:
This is the only change I'm unsure about. The raw data length is validated after decompression, so it should be guaranteed to be correct at this point. If it isn't, then something has gone wrong and surely we would rather crash than return None?
https://github.com/shssoichiro/oxipng/blob/f4e631bce773fc5ef67fae0f75a5fa4ea87f2110/src/png/mod.rs#L145-L148

Reply:
I've added an assert above for crashing. 
This condition is to make compiled code more efficient, because panicking branches inhibit optimizations and add code bloat.

Reply:
Ah, good point, thanks for the explanation. I might just add a comment in there for my own benefit 🙂